akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSinkSpec.scala

```diff
@@ -0,0 +1,145 @@
+
+  def expectSuccess[T](f: Future[T], value: T) =
+    Await.result(f, timeout) should be(value)
+
+  "InputStreamSink" must {
+    "read bytes from InputStream" in assertAllStagesStopped {
+      val (probe, (inputStream, f)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      probe.sendNext(byteString)
+      val arr = new Array[Byte](3)
+      inputStream.read(arr)
+      assert(arr === "abc".toArray)
+
+      probe.sendComplete()
+      expectSuccess(f, 3)
+    }
+
+    "block read until get requested number of bytes from upstream" in assertAllStagesStopped {
+      val (probe, (inputStream, _)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      val arr = new Array[Byte](3)
+      val f = Future(inputStream.read(arr))
+
+      expectBlocked(f)
+      probe.sendNext(byteString)
+      expectSuccess(f, 3)
+
+      probe.sendComplete()
+    }
+
+    "fill up buffer by default" in assertAllStagesStopped {
+      import system.dispatcher
+      val (probe, (inputStream, _)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      probe.sendNext(ByteString("abc"))
+      probe.sendNext(ByteString("def"))
+
+      val arr1 = new Array[Byte](3)
+      val arr2 = new Array[Byte](3)
+      val f1 = Future(inputStream.read(arr1))
+      val f2 = Future(inputStream.read(arr2))
+      Await.result(f1, timeout) should be(3)
+      Await.result(f2, timeout) should be(3)
+
+      assert(arr1 === "abc".toArray)
+      assert(arr2 === "def".toArray)
+
+      probe.sendComplete()
+    }
+
+    "throw error when reactive stream is closed" in assertAllStagesStopped {
+      val (probe, (inputStream, _)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      probe.sendNext(byteString)
+
+      inputStream.close()
+      probe.expectCancellation()
+
+      val arr = new Array[Byte](3)
+      //this is the cause of dead letter
+      the[Exception] thrownBy inputStream.read(arr) shouldBe a[IOException]
+    }
+
+    "unblock read when upstream is completed" in assertAllStagesStopped {
+      val (probe, (inputStream, _)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      probe.sendNext(ByteString("a"))
+      val arr = new Array[Byte](3)
+      val f = Future(inputStream.read(arr))
+
+      expectBlocked(f)
+      probe.sendComplete()
+      expectSuccess(f, 1)
+      assert(arr === "a\u0000\u0000".toArray)
+    }
+
+    "throw IOException when read after stream is completed" in assertAllStagesStopped {
+      val (probe, (inputStream, f)) = TestSource.probe[ByteString].toMat(InputStreamSink())(Keep.both).run()
+
+      probe.sendNext(byteString)
+      val arr = new Array[Byte](3)
+      inputStream.read(arr)
+
+      probe.sendComplete()
+      //this is the cause of dead letter
+      the[Exception] thrownBy Await.result(Future(inputStream.read(arr)), timeout) shouldBe a[IOException]
```

Comment:
there should be another test case that fails the stream, in which case the IOException is the right behavior